### ISSUE TITLE
Question about Dual Seg Head

### ISSUE THREAD

**Issue by SpeedyGonzales949** [valeoai/TTYD#3]
Hi! Thank you so much for the excellent work!

I have a few questions regarding the adaptation process, particularly about the dataset mapping.

1. How do you adapt the final linear layer for source-only evaluation on the shared set? After argmax, you map the class indices to the shared set? 
2. What was the reasoning behind training on the full set of source classes rather than just the shared subset?
3. When fine-tuning the model for the new domain, do you train a new linear layer from scratch? I noticed there is a parameter called dual_seg_head.

PS: shared set, I am referring to the middle column presented in the Appendix in your paper. 

Thanks a lot for the help.

**Comment by BjoernMichele** [comment]
Dear @SpeedyGonzales949, 

Thank you very much for your interest in our work. 

Regarding the questions: 
1. Yes, you describe it in the righ way.  In the evaluation (see here: https://github.dev/valeoai/TTYD/blob/0aa8126d196396a3eb3a29bac62d83eec7482b72/utils_source_free/utils.py#L254) we call the ``prediction changer'' function, which maps to the target classes (in this class mapping we also take into account that for some classes there is no mapping to the target).
2. There are several reasons: It is firstly much more computationally efficient, as with this you have for each source dataset exactly 1 trained source model (if you train with the subset, you have to retrain the source model for each target datset). Besides being computational more efficient, in my personal opinion, this is also closer to the idea of the source-free setting, where at the train time of the source-only model you not necessarily know which will be the target dataset. 
3. We have evaluated different approaches. See Table 6a) (https://arxiv.org/pdf/2409.04409), it is deep down in the appendix. However, if the classification layer was adapted(what is not what we finally went with (we went with :BN→ Lin, w. bias in the table)), we always started from the classification layer that we have obtained from the source model (so not intialised from scratch). The name dual_seg_head, is actually just a leftover from the SALUDA codebase (my previous project) that I started the code from. Sorry for the confusion. 


For now I close the issue, but feel very free to reopen if there are any further questions. 

With best regards, 
Bjoern

**Comment by SpeedyGonzales949** [comment]
Thanks a lot for the clear explanation.

**Comment by SpeedyGonzales949** [comment]
I have a quick follow-up question regarding a detail in the mapping process.

I understand how class mapping is handled during validation, that part is straightforward. However, I'm unclear about how the mapping of class probabilities is handled for the CrossEntropyLoss optimization.

- Since CrossEntropyLoss only requires the predicted logits for the ground truth class, how exactly is the mapping from source to target done in this context? Do you know if this has a special effect on the network ? Because now you have multiple neurons for the same class from the target domain.

- How do you convert the logits from shape [1, N_source] to [1, N_target] ? Do you first find the class with the highest probability in the source space, map that to a target class (if it exists), and reshape the output accordingly?

- Also, what happens if a source class has no valid mapping to the target domain? Do you treat those cases as ignored by setting the label to -1 (i.e., using ignore_index=-1)?

**Comment by BjoernMichele** [comment]
Sorry for my delay again. 

For question 1 and 2: 
- During training, we simply sum up all the classes that are merged into one class (one could think of doing the summing before or after an additional softmax. It can make sense if a lot of classes are merged that one uses a softmax  before to avoid getting too many negative logits from the other classes that are merged, if only one is detected). 

For question 3: 
- Yes, we set it to the ignore class / or 0 in the class probabilities.

**Comment by SpeedyGonzales949** [comment]
Hi, it's me again! I noticed that your paper, and several others, do not showcase the domain adaptation process of SemanticKitti -> NuScenes. Is there a specific reason why you decided against it ? Same reason for Waymo -> NuScenes?

Thanks for the help.

**Comment by BjoernMichele** [comment]
Dear @SpeedyGonzales949 ,

we did it mainly to follow the literature and the existing benchmarks. One technical motivation could be that going from 32 to 64 beams might be interesting as maybe newer technical sensors might also be denser, but also that going from 64 to 32 for example could be simulated by dropping beams (whereas extrapolation might be more difficult).

Best, 
Björn
